File: src/undirected-graph/undirectedGraph.py
# This module implements graph datastructure using python lists
# The idea is to get a high level understanding of graph data
# structure and the associated operations

from collections import deque

class Vertex:
	def __init__(self, nodeid):
		self.nodeid = nodeid
		self.neighbours = list()

	def GetNeighbours(self):
		return self.neighbours

class Graph:
	def __init__(self, size, is_directed = False):
		self.no_of_vertices = size
		self.no_of_edges = 0
		self.vertices = list()
		self.is_directed = is_directed
		for i in xrange(size):
			self.vertices.append(Vertex(i))

	def AddEdge(self, src, dest):
		self.vertices[src].neighbours.append(self.vertices[dest])
		if not self.is_directed:
			self.vertices[dest].neighbours.append(self.vertices[src])
		self.no_of_edges += 1

	def BreadthFirstSearch(self, u):
		explored_nodes = set([self.vertices[u]])
		candidate_nodes = deque([self.vertices[u]])
		while candidate_nodes:
			candidate = candidate_nodes.pop() # FIFO
			for neighbour in candidate.GetNeighbours():
				if neighbour not in explored_nodes:
					print ("explored %d" % neighbour.nodeid)
					explored_nodes.add(neighbour)
					candidate_nodes.appendleft(neighbour) 
		return explored_nodes

	def DepthFirstSearch(self, u):
		explored_nodes = set([self.vertices[u]])
		candidate_nodes = deque([self.vertices[u]])
		while candidate_nodes:
			candidate = candidate_nodes.popleft() # LIFO
			for neighbour in candidate.GetNeighbours():
				if neighbour not in explored_nodes:
					print ("explored %d" % neighbour.nodeid)
					explored_nodes.add(neighbour)
					candidate_nodes.appendleft(neighbour) 
		return explored_nodes

File: src/undirected-graph/undirected-graph.go
//Package undirected-graph provides graph data structure representation
//using adjacency lists. For simple purpose of learning, the vertices are
//assumed to be integers.
package undirected_graph

import ("container/list"
        "fmt")

type Vertex struct{
  id int
  neighbours *list.List
}

type Graph struct{
   vertices []Vertex
   no_of_vertices int
   no_of_edges int
}

func NewGraph(size int) *Graph{
  g := new(Graph)
  g.vertices = make([]Vertex, size)
  g.no_of_vertices = size
  //Initialize the graph
  for v := 0 ; v < size ; v++ {
    g.vertices[v].id = v
    g.vertices[v].neighbours = list.New()
  }
  return g
}

func (g *Graph) AddEdge(u, v int) {

  g.vertices[u].neighbours.PushBack(v)
  g.vertices[v].neighbours.PushBack(u)
  g.no_of_edges += 1
}

func (g *Graph) GetNeighbours(u int) *list.List {
   if (g.no_of_vertices == 0){
     return nil
   } else {
      neighbour_list := g.vertices[u].neighbours 
      return neighbours_list
}
}

//BFS method implements breadth first search on the graph using FIFO queue.
//This is a dummy implementation which just prints all the nodes as they
//are traversed.

func (g *Graph) BFS(u int) {

  //First add the given vertex to a queue
  main_queue := list.New()
  main_queue.PushBack(g.vertices[u])

  for e := main_queue.Front() ; e != nil ; e = e.Next() {

    
}
}
   

  

